godotproject/Player.gd

```diff
@@ -135,43 +240,84 @@ func do_movement(delta):
+		# Update our stored floor state.
+		on_floor_last_frame = is_on_floor()
+
 		# Count down the hop sound timer, ensuring it never goes below 0.
 		hop_sound_timer = max(hop_sound_timer - delta, 0)
-		# Play sound effect when timer has expired and frog is walking on floor.
-		if hop_sound_timer <= 0 && is_on_floor() && !is_zero(user_direction.x) && !is_zero(velocity.x):
+		# Play sound effect when timer has expired and frog is walking on floor
+		# (only if user is pointing in the same direction that frog is going)
+		if hop_sound_timer <= 0 && on_floor_last_frame && !is_zero(user_direction.x) && sign(velocity.x) == sign(user_direction.x):
 			# Calculate new timer value, based on period as well as random variance.
 			# This slight variance will hopefully make it sound more natural.
 			# Let P and V be the period and variance, the minimum timer value will be P and the max will be P*(1+V)
 			# In practice, with (P,V) = (0.2,0.1), the min will be 0.2 and the max will be 0.22
 			hop_sound_timer = hop_sound_timer_period * (1 + rand_range(0, hop_sound_timer_period_variance))
 			# Play sound effect.
 			$HopSoundPlayer.play(0)
-	pass
+
+func handle_jump_starting():
+	# Adjust jump start related timers used to permit users to jump a short bit
+	# after they technically left the ground or pressed the jump key.
+	# This is done in many games to prevent the user from feeling like their jump was
+	# "stolen" when they actually pressed the button after leaving the ground.
+	# Also allows them to press the jump button just before hitting the ground and
+	# not have their jump eaten. Technically this allows for the user to buffer a
+	# frame-perfect jump which preserves all their horizontal speed, but we can block
+	# this if users prove to be able to perform these jumps very easily.
+	if jump_pressed:
+		$JumpButtonPressLeniencyTimer.start(jump_leniency_time)
+	if on_floor_last_frame:
+		$JumpOnGroundLeniencyTimer.start(jump_leniency_time)
+	
+	# Check if we should jump - if we hit the jump button and were on solid ground recently
+	if $JumpButtonPressLeniencyTimer.time_left > 0 && $JumpOnGroundLeniencyTimer.time_left > 0:
+		# Reset both timers so we don't jump again next update.
+		$JumpButtonPressLeniencyTimer.stop()
+		$JumpOnGroundLeniencyTimer.stop()
+		# Set jump timers to cause frog to jump. The actual jumping motion is handled in do_movement
+		# on the next _physics_process call.
+		jump_frame_countdown = jump_length
+		jump_shorthop_countdown = shorthop_length
+		$JumpSoundPlayer.play(0)
 
 func update_anim():
+	# If the frog is shooting out the tongue, don't allow them to change
+	# which direction they're facing.
 	var facing_lock = $PlayerTongue.shooting
 	if !facing_lock:
-		if abs(user_direction.x) > 0.01:
+		# If the user is inputting left or right on the stick, face in that
+		# direction - otherwise, we will remain facing the direction that
+		# we were before.
+		if !is_zero(user_direction.x):
```

Comment:
is_zero(x) is a function that computes abs(x) > 0.01
Checking if a number is very close to zero is often better than checking for strict equality with zero, since it helps avoid floating point precision issues.
I factored it out since I used it frequently in the code.

Reply:
we can expand this to compare any float value, not just 0 within a margin. 
float_compare(x, y, alpha):
   if abs(x - y) > alpha:
      return false
   return true